### PR TITLE
Import patch from Alexander K. to make docking library fully void-safe.

### PR DESCRIPTION
**Pull request by jocelyn**
However this patch includes a small breaking changes (SD_CONTENT creation is now taking among other args, the docking manager value).
But the callers code is simple to fix, so probably acceptable.

### CODE REVIEW

File: Src/library/docking/implementation/configuration/sd_open_config_mediator.e [hunk 1]
```diff
@@ -25,8 +25,8 @@ feature {NONE} -- Initialization
 			internal_docking_manager := a_docking_manager
 
 			create cleaner.make (a_docking_manager)
-			create editor_helper
-			editor_helper.set_open_config_mediator (Current)
+			create editor_helper.make (Current)
+--			editor_helper.set_open_config_mediator (Current)
```
Comment (manuseiffel) [inline Src/library/docking/implementation/configuration/sd_open_config_mediator.e:29]: Comment to be removed.

File: Src/library/docking/implementation/configuration/sd_open_config_mediator.e [hunk 2]
```diff
@@ -67,7 +67,6 @@ feature -- Open inner container data
 			internal_docking_manager.command.recover_normal_state
 
 			internal_docking_manager.command.resize (True)
-			internal_docking_manager.property.set_is_opening_config (True)
```
Comment (manuseiffel) [inline Src/library/docking/implementation/configuration/sd_open_config_mediator.e:70]: Why was this moved and not the `resize` one?

File: Src/library/docking/implementation/configuration/sd_open_config_mediator.e [hunk 3]
```diff
@@ -421,14 +418,15 @@ feature {NONE} -- Implementation
...
 					end
 
 					if attached l_top_split_area.second as l_second then
 						l_right_zones := internal_docking_manager.query.all_zones_in_widget (l_second)
 					else
 						check False end -- Implied by docking widget must not has empty child
+						create l_right_zones.make (0)
 					end
-					check l_left_zones /= Void and l_right_zones /= Void end -- Implied by previous if clauses
 					if is_all_tools (l_left_zones) or  is_all_tools (l_right_zones) then
```
Comment (manuseiffel) [inline Src/library/docking/implementation/configuration/sd_open_config_mediator.e:430]: Couldn't we make `is_all_tools` return false when argument is Void? That would save us from having to create them when something goes wrong.

File: Src/library/docking/implementation/configuration/sd_open_config_mediator.e [hunk 4]
```diff
@@ -564,14 +562,21 @@ feature {NONE} -- Implementation
...
+						%to avoid potential void-safety violations if not all attached attributes are properly set.")
+				end
 				create l_internal
 				if attached a_config_data.state as l_state then
 					l_type_id := l_internal.dynamic_type_from_string (l_state)
 				end
 
 				check a_type_exist: l_type_id /= -1 end
-				if attached {SD_STATE} l_internal.new_instance_of (l_type_id) as l_state then
+				if attached {SD_STATE_WITH_CONTENT} l_internal.new_instance_of (l_type_id) as l_state then
```
Comment (manuseiffel) [inline Src/library/docking/implementation/configuration/sd_open_config_mediator.e:579]: Was this a bug before to use SD_STATE?

File: Src/library/docking/implementation/configuration/sd_open_config_mediator.e [hunk 5]
```diff
@@ -809,30 +810,33 @@ feature {NONE} -- Implementation
...
 
-				l_tool_bar.float (l_data.screen_x, l_data.screen_y, l_data.is_visible)
-				l_content.set_visible (l_data.is_visible)
+					l_tool_bar.float (l_data.screen_x, l_data.screen_y, l_data.is_visible)
+					l_content.set_visible (l_data.is_visible)
 
-				if l_state /= Void and then attached l_state.floating_group_info as l_info then
-					l_tool_bar.attached_floating_tool_bar.assistant.position_groups (l_info)
+					if
+						attached l_state and then
```
Comment (manuseiffel) [inline Src/library/docking/implementation/configuration/sd_open_config_mediator.e:834]: We should use `l_state /= Void` like above

File: Src/library/docking/implementation/configuration/sd_open_config_mediator.e [hunk 6]
```diff
@@ -809,30 +810,33 @@ feature {NONE} -- Implementation
 			loop
 				l_data := a_tool_bar_data.item
 				check is_floating_tool_bar_data: l_data.is_floating end
-				l_title := l_data.title
-				check l_title /= Void end -- Implied by check `is_floating_tool_bar_data'
-				l_content := l_tool_bar_manager.content_by_title (l_title)
+				if attached l_data.title as l_title then
```
Comment (manuseiffel) [inline Src/library/docking/implementation/configuration/sd_open_config_mediator.e:813]: Wouldn't it better to have `content_by_title` return Void if no good title is provided? I looked at the code and we would end up in a crash if ever in the loop we cannot set Result.

File: Src/library/docking/implementation/controls/notebook/sd_notebook_tab.e [hunk 7]
```diff
@@ -225,45 +227,51 @@ feature -- Query
...
 			parented: parent /= Void
-		local
-			l_parent: like parent
 		do
-			l_parent := parent
-			check l_parent /= Void end -- Implied by precondition `parented'
-			Result := x + l_parent.screen_x
+			if attached parent as l_parent then
+				Result := x + l_parent.screen_x
+			else
```
Comment (manuseiffel) [inline Src/library/docking/implementation/controls/notebook/sd_notebook_tab.e:255]: No need for the check

File: Src/library/docking/implementation/controls/notebook/sd_notebook_tab.e [hunk 8]
```diff
@@ -225,45 +227,51 @@ feature -- Query
...
 			parented: parent /= Void
-		local
-			l_parent: like parent
 		do
-			l_parent := parent
-			check l_parent /= Void end -- Implied by precondition `parented'
-			Result := l_parent.screen_y
+			if attached parent as l_parent then
+				Result := l_parent.screen_y
+			else
```
Comment (manuseiffel) [inline Src/library/docking/implementation/controls/notebook/sd_notebook_tab.e:270]: No need for the check

File: Src/library/docking/implementation/controls/sd_constants_imp.e [hunk 9]
```diff
@@ -12,21 +12,21 @@ feature {NONE} -- Initialization
...
-					file.open_read
-					file.readstream (file.count)
-					file.close
-					parse_file_contents (file.laststring)
-				end
-				initialized_cell.put (True)
-			end
-		ensure
-			initialized: constants_initialized
+--			if not constants_initialized then
```
Comment (manuseiffel) [inline Src/library/docking/implementation/controls/sd_constants_imp.e:18]: Why are this and below code commented?

File: Src/library/docking/implementation/controls/tool_bar/sd_tool_bar_popup_button.e [hunk 10]
```diff
@@ -49,16 +49,20 @@ feature -- Command
...
-	set_popup_widget (a_widget: like popup_widget)
+	set_popup_widget (a_widget: attached like popup_widget)
 			-- Set `popup_widget' with `a_widget'
 		require
 			not_void: a_widget /= Void
 		do
 			popup_widget := a_widget
 			popup.wipe_out
-			popup.extend (a_widget)
+			if attached a_widget then
```
Comment (manuseiffel) [inline Src/library/docking/implementation/controls/tool_bar/sd_tool_bar_popup_button.e:59]: if `a_widget` is attached, no need for the object test

File: Src/library/docking/implementation/docking/sd_docking_manager_command.e [hunk 11]
```diff
@@ -17,31 +17,29 @@ inherit
...
 
 feature {NONE}  -- Initlization
 
-	make
-			-- Creation method
+	make (a_docking_manager: SD_DOCKING_MANAGER)
+			-- Associate new object with `a_docking_manager'.
 		do
+			docking_manager := a_docking_manager
 			create locked_windows.make (10)
```
Comment (manuseiffel) [inline Src/library/docking/implementation/docking/sd_docking_manager_command.e:30]: Looks to me that we should be using an ARRAYED_LIST for `locked_windows`. Not required now, but it would slightly simplify the code now.

File: Src/library/docking/implementation/docking/sd_docking_manager_query.e [hunk 12]
```diff
@@ -13,28 +13,38 @@ inherit
...
+			when {SD_ENUMERATION}.bottom then
+				Result := docking_manager.internal_auto_hide_panel_bottom
+			when {SD_ENUMERATION}.top then
+				Result := docking_manager.internal_auto_hide_panel_top
+			when {SD_ENUMERATION}.left then
+				Result := docking_manager.internal_auto_hide_panel_left
+			when {SD_ENUMERATION}.right then
+				Result := docking_manager.internal_auto_hide_panel_right
+			end
+			if not attached Result then
```
Comment (manuseiffel) [inline Src/library/docking/implementation/docking/sd_docking_manager_query.e:45]: No need for the check since there is no else clause in the inspect, we are guaranteed that Result is attached if all the above assignments have this post-condition.

**Comment by jocelyn** [comment]
already merged
